ls1intum/Artemis
# `Programming exercises`: Add git-diff generation between template and solution repositories

### Checklist
#### General
#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context


These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java.
In order to create code hints for behavioral test cases, we will need the difference between the template and solution repositories in addition to a test coverage. This is required, so we do not generate code hints for code that is already given in the template.

### Description

This PR introduces the generation of such a difference utilizing the git-diff tool.
It uses JGit to generate the git-diff between the template and solution repositories and then converts it to a single ProgrammingExerciseGitDiffReport with one ProgrammingExerciseGitDiffEntry per change.
In order to save space in the database these diff entries do *not* contain the actual code changes. They only contain the file paths, start lines of the code blocks and the length of the code blocks (in lines).
ProgrammingExerciseFullGitDiffReportDTO and ProgrammingExerciseFullGitDiffEntryDTO are derived from these and *do* contain the actual code blocks. They are created from the light diff-report in ProgrammingExerciseGitDiffService.

The light git-diff is created every time the template or solution repositories get updated. The full git-diff is created only when needed by the client. (currently only when the client requests it with a button in the detail view) This is also the reason they are merely DTOs.

**UML Diagram:**

<img width="600" alt="grafik" src="https://user-images.githubusercontent.com/18218285/156144396-5cda8827-3344-4b3a-8870-cc1ec017b8d7.png">

**Example:**
The git-diff output
```
--- src/test.java
+++ src/test.java
@@ -1,1 +1,3 @@
- // TODO: Add class
+ public class Test {
+     // TODO: Implement class
+ }
```
would result in the following (light) DiffEntry
```
previousFilePath: "src/test.java",
filePath: "src/test.java",
previousStartLine: 1,
startLine: 1,
previousLineCount: 1,
lineCount: 3
```
and the following FullDiffEntry
```
previousFilePath: "src/test.java",
filePath: "src/test.java",
previousLine: 1,
line: 1,
previousCode: "// TODO: Add class",
code: """
public class Test {
    // TODO: Implement class
}
"""
```

I have also added a client component to display the GitDiffReport to the user in a GitHub-like way.

### Steps for Testing

Prerequisites:
- 1 Editor or higher
- 1 Programming Exercise

1. Log in to Artemis
2. Navigate to the instructor detail page of the exercise
3. Change at least the exercise solution (Not necessary if you newly created the exercise in this PR)
4. Wait until the build is finished
5. Click on the new git-diff button
6. Check that the displayed git-diff matches the changes you did between template and solution

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





| File | Branches % | Lines % |
| --- | --- | --- |
| ProgrammingExerciseGitDiffEntry | 100 | 94 |
| ProgrammingExerciseGitDiffReport | 100 | 100 |
| ProgrammingExerciseGitDiffReportService | 73 | 90 |
| ProgrammingExerciseGitDiffReportResource | 50 | 89 |
| git-diff-entry.component.ts | 76.47 | 100 |
| git-diff-modal.component.ts | 75 | 66.66 |
| git-diff-report.component.ts | 88.88 | 100 |

### Screenshots


#### The new button:
<img width="412" alt="grafik" src="https://user-images.githubusercontent.com/18218285/156145414-c186330c-163c-4a3a-8331-085765d9c5f3.png">


#### The modal with the git-diff:
<img width="1180" alt="grafik" src="https://user-images.githubusercontent.com/18218285/156143078-4e7f92aa-e509-4765-91d2-312db1097000.png">
